lib/temporal/worker.rb

```diff
@@ -0,0 +1,72 @@
+require 'async'
+require 'temporal/bridge'
+require 'temporal/data_converter'
+require 'temporal/failure_converter'
+require 'temporal/payload_converter'
+require 'temporal/runtime'
+require 'temporal/worker/activity'
+require 'temporal/worker/thread_pool_executor'
+
+module Temporal
+  class Worker
+    # TODO: Add worker interceptors
+    def initialize(
+      connection,
+      namespace,
+      task_queue,
+      payload_converter: Temporal::PayloadConverter::DEFAULT,
+      payload_codecs: [],
+      failure_converter: Temporal::FailureConverter::DEFAULT,
+      activity_executor: nil,
+      max_concurrent_activities: 100
+    )
+      @running = false
+      @mutex = Mutex.new
+      @runtime = Temporal::Runtime.instance
+      activity_executor ||= ThreadPoolExecutor.new(max_concurrent_activities)
+      data_converter = DataConverter.new(
+        payload_converter: payload_converter,
+        payload_codecs: payload_codecs,
+        failure_converter: failure_converter,
+      )
+      core_worker = Temporal::Bridge::Worker.create(
+        @runtime.core_runtime,
+        connection.core_connection,
+        namespace,
+        task_queue,
+      )
+      @activity_worker = Worker::Activity.new(core_worker, data_converter, activity_executor)
+      @workflow_worker = nil
+    end
+
+    def run
+      Async { |task| start(task) }
+    end
+
+    def start(reactor = nil)
+      mutex.synchronize do
+        raise 'Worker is already running' if running?
+
+        @running = true
+      end
+
+      runtime.ensure_callback_loop
+      reactor ||= runtime.reactor
+      reactor.async { |task| activity_worker.run(task) } if activity_worker
+      reactor.async { |task| workflow_worker.run(task) } if workflow_worker
+    end
+
+    def shutdown
+      activity_worker&.shutdown
+      workflow_worker&.shutdown
+    end
+
+    private
+
+    attr_reader :mutex, :runtime, :activity_worker, :workflow_worker
+
+    def running?
+      @running
```

Comment:
Can you confirm that Ruby variable reads are atomic and thread safe (even though obv writes are not)? Some interpreters I know use multiple non-locked instructions to read variables.

Reply:
I believe they are because of the GVL (global VM lock), which only allows a single CPU-bound instruction to be executed at the same time anyways (across however many Ruby threads). But I'm only accessing it via a mutex anyways on line 47 anyways

Reply:
Makes sense

> But I'm only accessing it via a mutex anyways on line 47 anyways

Also accessed here by any user calling `running?` I assume. But no prob if two ruby lines can't be executed at the same time.

Reply:
> But no prob if two ruby lines can't be executed at the same time.

Yep, which is also true for Python as far as I know

Reply:
Correct (and we take advantage of that in several places)